bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelServerRoutingV2.java

```diff
@@ -0,0 +1,101 @@
+import static org.eclipse.emfcloud.modelserver.emf.common.util.ContextResponse.missingParameter;
+
+import java.util.Optional;
+
+import org.apache.log4j.Logger;
+import org.eclipse.emfcloud.modelserver.common.ModelServerPathsV2;
+
+import com.google.inject.Inject;
+
+import io.javalin.Javalin;
+import io.javalin.http.Context;
+import io.javalin.websocket.WsConnectContext;
+import io.javalin.websocket.WsHandler;
+
+public class ModelServerRoutingV2 extends AbstractModelServerRouting {
+   protected static final Logger LOG = Logger.getLogger(ModelServerRoutingV2.class.getSimpleName());
+
+   protected final TransactionController transactionController;
+
+   @Inject
+   public ModelServerRoutingV2(final Javalin javalin, final ModelResourceManager resourceManager,
+      final ModelController modelController, final SchemaController schemaController,
+      final ServerController serverController, final SessionController sessionController,
+      final TransactionController transactionController) {
+
+      super(javalin, resourceManager, modelController, schemaController, serverController, sessionController);
+
+      this.transactionController = transactionController;
+   }
+
+   @Override
+   protected String getBasePath() { return ModelServerPathsV2.BASE_PATH; }
+
+   @Override
+   protected void apiEndpoints() {
+      super.apiEndpoints();
+
+      put(ModelServerPathsV2.MODEL_BASE_PATH, this::setModel); // Was PATCH in V1
+
+      // PATCH /edit removed in V2; replaced with PATCH /models TODO
+      patch(ModelServerPathsV2.MODEL_BASE_PATH, this::executeCommand); // Was PATCH /edit in V1
+
+      post(ModelServerPathsV2.TRANSACTION, this::createTransaction);
+      ws(ModelServerPathsV2.TRANSACTION + "/:id", this::openTransaction);
+   }
+
+   @Override
+   protected void onSubscriptionConnect(final WsConnectContext ctx) {
+      Optional<String> modelUri = getResolvedFileUri(ctx.queryParamMap(), MODEL_URI);
+      if (modelUri.isEmpty()) {
+         missingParameter(ctx, MODEL_URI);
+         ctx.session.close();
+         return;
+      }
+
+      if (!sessionController.subscribeV2(ctx, modelUri.get())) {
+         error(ctx, "Cannot subscribe to '%s': modeluri is not a valid model resource", modelUri.get());
+         ctx.session.close();
+      }
+   }
+
+   protected void createTransaction(final Context ctx) {
+      getResolvedFileUri(ctx, MODEL_URI).ifPresentOrElse(
+         param -> transactionController.create(ctx, param),
+         () -> missingParameter(ctx, MODEL_URI));
+   }
+
+   protected void openTransaction(final WsHandler wsHandler) {
+      wsHandler.onConnect(transactionController::onOpen);
+      wsHandler.onClose(transactionController::onClose);
+      wsHandler.onError(transactionController::onError);
+      wsHandler.onMessage(transactionController::onMessage);
+   }
+
+   @Override
+   protected void executeCommand(final Context ctx) {
+      // TODO: this was rewired from PATCH /edit to PATCH /models
```

Comment:
Create issue for this?

Reply:
We should probably consider this is the context of this PR, as this might cause unexpected behavior for V1 (Breaking legacy behavior).